roadmap/implementers-guide/src/node/dispute/votesdb.md

```diff
@@ -0,0 +1,139 @@
+# Votes DB
+
+Responsible module for tracking all kind of dispute votes by all validators
+for a fixed number of sessions.
+
+
+## Remarks
+
+Votes must be persisted. Disputes might happen long after
+there was a dispute, as such just keeping things in memory
+is thus insufficient, as every blip (OOM, bug, maintanance)
+could cause the node to forget the state.
+
+`UnsignedTransaction`s are OK to be used, since the inner
+element `CommittedCanddidateReceipt` is verifiable.
+
+## ToDo
+
+* Missing Definition of Governance Mode.
+* Is a `PoV` enough or should the whole `CommittedCandidateReceipt` be included?
+* Currently [`fn mark_bad()`](https://github.com/paritytech/substrate/pull/6301/files#diff-8faeb5c685a8fdff428c5ec6d9102fd59e127ff69762d43045cd38e586db5559R60-R64) does not persist data.
+
+## IO
+
+Inputs:
+
+* `VotesDbMessage::`
+
+Outputs:
+
+* `DisputeConclusionMessage::Detection`
+* `DisputeConclusionMessage::Resolution`
+* `ChainApiMessage::Blacklist`
+
+## Messages
+
+```rust
+enum VotesDbMessage {
+    /// Allow querying all `Hash`es queried by a particular validator.
+    QueryValidatorVotes {
+        /// Validator indentification.
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        response: ResponseChannel<Vec<Hash>>,
+    },
+
+    /// Store a vote for a particular dispute
+    StoreVote{
+        /// Unique validator indentification
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        /// Vote.
+        vote: Vote,
+        /// Attestation.
+        attestation: ValidityAttestation,
+    },
+}
+```
+
+```rust
+enum DisputeMessage {
+    /// A dispute is detected
+    Detection {
+        /// unique validator indentification
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        /// The attestation.
+        attestation: ValidityAttestation,
+        /// response channel
+        response: ResponseChannel<Vec<Hash>>,
+    },
+    /// Concluded a dispute with the following resolution
+    Resolution {
+        /// Resolution of the block can either be,
+        /// the block being valid or not.
+        valid: bool,
+    }
+}
+```
+
+## Helper Structs
+
+```rust
+/// Snapshot of the current vote state and recorded votes
+struct Snapshot {
```

Comment:
Where are these `Snapshot`s stored? If they're all votes on a candidate, then that shouldn't cross session boundaries. And it should be possible to have more types of votes than just `ValidityAttestation` - there's also `Approval`, `DisputePositive`, and `DisputeNegative`